torchmultimodal/models/gpt.py

```diff
@@ -28,12 +28,221 @@ class TransformerLayerOutput(NamedTuple):
+        use_cache (bool, optional): If ``True``, caches past key/value tensors for faster decoding. If ``False``,
+            recomputes key and value for each decoding step. Defaults to ``False``.
+        causal (bool. optional): If ``True``, use causal attention. Defaults to ``False``.
+        right_shift (bool): If ``True``, shifts the embedding vectors to the right and prepends it with start of
+            sentence token. Defaults to ``False``. This option is disregarded during training mode
+        return_attn_weights (bool, optional): If ``True``, returns attention probabilities of each transformer
+            layer. Defaults to ``False``.
+        return_hidden_states (bool): If ``True``, returns the embeddings of each transformer layer. Defaults to
+            ``False``.
+
+    Raises:
+        AttributeError: If input tokenizer does not implement methods ``encode`` and ``decode`` or output
+            tokenizer does not implement methods ``encode``, ``decode`` and ``lookup``.
+    """
+
+    def __init__(
+        self,
+        d_model: int,
+        num_tokens: int,
+        in_tokenizer: nn.Module,
+        out_tokenizer: nn.Module,
+        mm_decoder: nn.Module,
+    ) -> None:
+        super().__init__()
+        if not all(
+            [hasattr(in_tokenizer, attr_name) for attr_name in ["encode", "decode"]]
+        ):
+            raise AttributeError(
+                "Input modality tokenizer must have methods 'encode', 'decode' and 'lookup'."
+            )
+
+        if not all(
+            [
+                hasattr(out_tokenizer, attr_name)
+                for attr_name in ["encode", "decode", "lookup"]
+            ]
+        ):
+            raise AttributeError(
+                "Output modality tokenizer must have methods 'encode', 'decode' and 'lookup'."
+            )
+
+        self.d_model = d_model
+        self.num_tokens = num_tokens
+        self.in_tokenizer = in_tokenizer
+        self.out_tokenizer = out_tokenizer
+        self.mm_decoder = mm_decoder
+        self.norm = nn.LayerNorm(d_model)
+        self.to_logit = nn.Linear(d_model, num_tokens, bias=False)
+        # This will give us equal probabilities after the soft max layer initially to avoid biasing
+        # towards any particular prediction category
+        self.to_logit.weight.data.copy_(torch.zeros(self.num_tokens, self.d_model))
+
+    def forward(
+        self,
+        in_modality: Optional[Tensor] = None,
+        out_modality: Optional[Tensor] = None,
+        in_pos_ids: Optional[Tensor] = None,
+        out_pos_ids: Optional[Tensor] = None,
+        attn_mask: Optional[Tensor] = None,
+        head_mask: Optional[Tensor] = None,
+        logits_mask: Optional[Tensor] = None,
+        use_cache: bool = False,
+        causal: bool = False,
+        right_shift: bool = False,
+        return_attn_weights: bool = False,
+        return_hidden_states: bool = False,
+    ) -> MultimodalGPTOutput:
```

Comment:
>During training:
- encode converts in- and out-modality data to quantized embeddings
- forward takes the quantized embeddings (concatenated) for training

I will move `encode` and add concatenation step to `forward`.

For generation, `forward` is not used as we have to use those components in separation except for `MultimodalTransformerDecoder` (that operates at token level).

Reply:
>I will move encode and add concatenation step to forward.

Ended up not doing this change: `encode` returns both the quantized embeddings and tokens. The latter are in fact "labels" (or ground truth) during training of the GPT decoder. Since we don't want to compute loss inside `forward`, I will keep `encode` and `forward` separate and line them up in the lightning module.